src/main/kotlin/eu/gillissen/rider/usersecrets/OpenUserSecretsAction.kt

```diff
@@ -2,75 +2,107 @@ package eu.gillissen.rider.usersecrets
-            actionEvent.presentation.isEnabled = false
+        val projectFile = actionEvent.getData(PlatformDataKeys.VIRTUAL_FILE)
+        if (projectFile == null || CsProjExtension != projectFile.extension) {
+            actionEvent.presentation.isEnabledAndVisible = false
             return
         }
 
-        val documentBuilder = DocumentBuilderFactory.newInstance().newDocumentBuilder()
-        val document = documentBuilder.parse(file.inputStream)
+        val document = DocumentBuilderFactory.newInstance()
+            .newDocumentBuilder()
+            .parse(projectFile.inputStream)
+
         document.documentElement.normalize()
-        val userSecretsIds = document.getElementsByTagName("UserSecretsId")
-        var textContent: String? = null
-        if (userSecretsIds.length > 0) {
-            textContent = userSecretsIds.item(0).textContent
+        val userSecretsIdNodes = document.getElementsByTagName(UserSecretsIdMsBuildProperty)
+        if (userSecretsIdNodes.length == 0) {
+            actionEvent.presentation.isEnabledAndVisible = false
+            return
         }
-        if (textContent.isNullOrEmpty()) {
+
+        if (userSecretsIdNodes.item(0).textContent.isNullOrEmpty()) {
             actionEvent.presentation.isEnabled = false
+            actionEvent.presentation.isVisible = true
             return
         }
 
-        chosenProject = project
-        id = textContent
-
         actionEvent.presentation.isEnabledAndVisible = true
     }
+
+    override fun actionPerformed(actionEvent: AnActionEvent) {
+
+        val project = CommonDataKeys.PROJECT.getData(actionEvent.dataContext) ?: return
+        if (project.isDefault) return
+
+        val projectFile = actionEvent.getData(PlatformDataKeys.VIRTUAL_FILE) ?: return
+        if (CsProjExtension != projectFile.extension) return
+
+        object : Task.Backgroundable(project, "Retrieving user secrets...", true, DEAF) {
+
+            override fun run(indicator: ProgressIndicator) {
+                val msBuildEvaluator = MSBuildEvaluator.getInstance(project)
+                val msBuildProperties = msBuildEvaluator
+                    .evaluateProperties(MSBuildEvaluator.PropertyRequest(projectFile.path, null, listOf(UserSecretsIdMsBuildProperty)))
+                    .blockingGet(1, TimeUnit.MINUTES)
+                    ?: return
+
+                val secretsId = msBuildProperties[UserSecretsIdMsBuildProperty] ?: return
+
+                ApplicationManager.getApplication().invokeLaterOnWriteThread {
+                    val secretsDirectoryRoot = getSecretsDirectoryRoot()
+                    val secretsDirectory = "$secretsDirectoryRoot${File.separatorChar}$secretsId"
+                    val secretsFile = File("$secretsDirectory${File.separatorChar}secrets.json")
+                    if (!secretsFile.exists()) {
+                        File(secretsDirectoryRoot).mkdirs()
+                        File(secretsDirectory).mkdirs()
```

Comment:
shouldn't just `File(secretsDirectory).mkdirs()` be sufficient here, as it creates all missing parent directories?

Reply:
You're right, updated.